include/shark/ObjectiveFunctions/NegativeLogLikelihood.h

```diff
@@ -104,56 +97,48 @@ class NegativeLogLikelihood : public AbstractObjectiveFunction< RealVector, doub
 		SIZE_CHECK(input.size() == numberOfVariables());
 		m_evaluationCounter++;
 		mep_model->setParameterVector(input);
-		derivative.resize(input.size());
-		derivative.clear();
 		
-		//compute partitioning on threads
-		std::size_t numBatches = m_data.numberOfBatches();
-		std::size_t numElements = m_data.numberOfElements();
-		std::size_t numThreads = std::min(SHARK_NUM_THREADS,numBatches);
-		//calculate optimal partitioning
-		std::size_t batchesPerThread = numBatches/numThreads;
-		std::size_t leftOver = numBatches - batchesPerThread*numThreads;
-		double error = 0;
-		double minProb = 1e-100;//numerical stability is only guaranteed for lower bounded probabilities
-		SHARK_PARALLEL_FOR(int ti = 0; ti < (int)numThreads; ++ti){//MSVC does not support unsigned integrals in paralll loops
-			std::size_t t = ti;
-			//~ //get start and end index of batch-range
-			std::size_t start = t*batchesPerThread+std::min(t,leftOver);
-			std::size_t end = (t+1)*batchesPerThread+std::min(t+1,leftOver);
-			
-			//calculate error and derivative of the current thread
-			FirstOrderDerivative threadDerivative(input.size(),0.0);
-			double threadError = 0;
-			boost::shared_ptr<State> state = mep_model->createState();
-			RealVector batchDerivative;
+
+		typedef std::pair<double,RealVector> result_type;
+		auto map = [this](RealMatrix const& batch){
+			double minProb = 1e-100;//numerical stability is only guaranteed for lower bounded probabilities
 			RealMatrix predictions;
-			for(std::size_t i  = start; i != end; ++i){
-				mep_model->eval(m_data.batch(i),predictions,*state);
-				SIZE_CHECK(predictions.size2() == 1);
-				threadError += sum(log(max(predictions,minProb)));
-				RealMatrix coeffs(predictions.size1(),predictions.size2(),0.0);
-				//the below handls numeric instabilities...
-				for(std::size_t j = 0; j != predictions.size1(); ++j){
-					for(std::size_t k = 0; k != predictions.size2(); ++k){
-						if(predictions(j,k) >= minProb){
-							coeffs(j,k) = 1.0/predictions(j,k);
-						}
+			boost::shared_ptr<State> state = mep_model->createState();
```

Comment:
Is there a specific reason for boost::shared_ptr instead of std::shared_ptr, or is this code just old?

Reply:
old code + we need boost::shared_ptr at some place because older boost::serialization versions can not cope with it. So instead of having different types of shared_ptr we pick the one that will not screw up in some cases.